daffodil-runtime1/src/main/scala/org/apache/daffodil/udf/UserDefinedFunctionService.scala

```diff
@@ -0,0 +1,185 @@
+  lazy val classUserDefinedFunction = classOf[UserDefinedFunction]
+  lazy val classJSerializable = classOf[java.io.Serializable]
+  lazy val currentClassPath = Misc.classPath.map(_.toString).mkString("\n")
+
+  val loader: ServiceLoader[UserDefinedFunctionProvider] = ServiceLoader.load(classOf[UserDefinedFunctionProvider])
+
+  loader.asScala.map { provider =>
+    val providerfc = provider.getUserDefinedFunctionClasses
+    lazy val providerClassName = provider.getClass.getName
+
+    if (providerfc == null || providerfc.isEmpty) {
+      val wstr = s"User Defined Function Provider ignored: ${providerClassName}. No User Defined Functions found."
+      log(LogLevel.Warning, wstr)
+    } else {
+      val missingAnnotations = providerfc.filterNot(_.isAnnotationPresent(classUserDefinedFunctionInfo))
+      val nonUserDefinedFunction = providerfc.filterNot(classUserDefinedFunction.isAssignableFrom(_))
+
+      if (!missingAnnotations.isEmpty) {
+        val wstr = s"User Defined Function Provider ignored: ${providerClassName}. Below must be annotated with ${classUserDefinedFunctionInfo.getName}" +
+          s"\n${missingAnnotations.map(_.getName).mkString("\t", "\n\t", "")}"
+        log(LogLevel.Warning, wstr)
+      }
+
+      if (!nonUserDefinedFunction.isEmpty) {
+        val wstr = s"User Defined Function Provider ignored: ${providerClassName}. Below must implement ${classUserDefinedFunction.getName}" +
+          s"\n${nonUserDefinedFunction.map(_.getName).mkString("\t", "\n\t", "")}"
+        log(LogLevel.Warning, wstr)
+      }
+
+      if (missingAnnotations.isEmpty && nonUserDefinedFunction.isEmpty) {
+        providerfc.map { fc =>
+          val fcClassName = fc.getName
+          val fInfo = fc.getDeclaredAnnotation(classUserDefinedFunctionInfo)
+          val fns = fInfo.namespaceURI()
+          val fname = fInfo.name()
+          if (List(fns, fname).exists { p => Option(p).isEmpty || p.trim.isEmpty }) {
+            val wstr = s"User Defined Function ignored: ${fcClassName} from provider ${providerClassName}." +
+              (if (Option(fns).isEmpty || fns.trim.isEmpty) " Annotation namespace field is empty or invalid." else "") +
+              (if (Option(fname).isEmpty || fname.trim.isEmpty) " Annotation namespace field is empty or invalid." else "")
+            log(LogLevel.Warning, wstr)
+          }
+          val key = s"$fns:$fname"
+          if (functionProviderLookup.contains(key)) {
+            val estr = s"User Defined Function ignored: ${fcClassName} from provider ${providerClassName}. Duplicate $key found."
+            log(LogLevel.Error, estr)
+          } else {
+            functionProviderLookup += (key -> provider)
+            functionClassesLookup += (key -> fc)
+          }
+        }
+      }
+    }
+  }
+
+  if (functionClassesLookup.isEmpty) {
+    log(LogLevel.Warning, "No user defined functions found. Check that UDF JARs are on classpath and that they are properly registerable by ServiceLoader.")
```

Comment:
Shouldn't this be an error? If the schema calls for functions for which there are no definitions this should be an SDE shouldn't it?  If the schema calls for no such functions, then this shouldn't be issuing a warning even.

Reply:
Steve explained it here: https://github.com/apache/incubator-daffodil/pull/273#discussion_r331499959

But essentially the error (which is reported in Expressions) is that the function called is not supported. We "warn" here that we found no UDFs, so they should check their classpath

Reply:
Do we have a way of knowing that the user intended to have a UDF. It seems like the common case is that the user intended to use a built-in function and typoed it (or mistakenly believed that Daffodil supported a standard XPATH function that it doesn't). In such a case, warning about classpath stuff is likely to confuse the user and lead them down a wrong path in debugging.

I agree with Steve's idea to list UDFs when debugging is enabled. An alternative is to unconditionally include "UDFs loaded: ..." In the standard function not found error message. That should let people who intended to use UDFs know there was a problem loading them, without confusing other users.

Reply:
We could maybe do some heuristics to determine if a user meant to use a UDF or not?

For example, if the namespace is not DFDL or FN then the user probably didn't intend to use a UDF and just typoed something or tried to use a function we dont' support?

My concern with "UDF's loaded" is that I can imagine cases where a user develops a large library of UDF's, which is going to lead to spam.

Another option might be to try to find the function the user meant, scanning for both UDF's and built-ins, and do a "did you mean on of these" kind of things. That's hard to get right too though.

Reply:
So functionClassLookup contains all the UDFs for all the UDFPs found by the ServiceLoader, so it'll only be empty when there were no UDFs stored for any of the UDFPs, or if there were no UDFPs found. So it's independent of any function call (with the exception of an unknown call triggering UDF(P) loading). 

This is intended to support the case, where UDFs are called in the schema, but maybe they didn't actually get put on the classpath (or any number of reasons that would cause the loader to not see them). This check shows them the classpath and tells them that no UDFs were loaded. We'd also get a "UDFP not found..." log error, a "UDF not found..." log error, and a "Error calling namespace:fn.." SDE. 
In the case of a typo with no UDFs loaded, we'd get the same.  

But in the case of a typo, with UDFs loaded, we wont get warning, and we'd get a "UDFP not found..." log error, a "UDF not found..." log error, and a "Error calling namespace:fn.." SDE

Is this undesired behavior?

Reply:
I would say we want to be consistent and have the same error message regardless if any UDF's are loaded. I think the easiest thing to do is to have the same "Unknown function" SDE when we do not recognnize a function, and then just have enough debug/info/warning logging that says when we ignore a UDF (and why) and which UDF's were successfully loaded that user can figure out why their particular UDF didn't work, if they intended to use a UDF. If they didn't intend to use a UDF (the most common case) then it's just a normal error and they'll probably recognize the typo.

Another thing we could potentialy do (so many options when it comes to good diagnostics), is keep track of which UDF's were ignored. If a user tries to use one of those, we can output a special error. Might be overcomplicating things though. I imagine it won't be common to be ignoring UDF's--they are pretty straitforward to create, and we have pretty good documentation/examples.

Reply:
Let's not try to make the perfect be the enemy of the good here. An error because the function is not defined is required, adding to that a warning when there are no UDFs defined is great.